File: Simple-NN-library/C++/src/functions/optimisers/adam.cpp
#include <functions/optimisers/adam.h>

using namespace std;
using namespace functions;
using namespace optimisers;
using namespace further_maths;


Adam::Adam(float learning_rate=0.001, float beta=0.9, float beta2=0.999, float decay=0){
  this.iterations=0;
  this.learning_rate=learning_rate;
  this.beta=beta;
  this.beta2=beta2;
  this.init_decay=decay;
  this.decay=decay;
  this.epsilon=epsilon();
}
Adam::~Adam(){

}

float[][2] Adam::update(float cost, float params[]=0){
  float updates[][2];
  gradients=make_array(0, (int)cost);
  float x=0;
  float it=x+1;
  float fix=(float)1-((float)1-this.beta)**it;
  float fix2=(float)1-((float)1-this.beta2)**it;
  float learning_rate_temp=this.learning_rate*(sqrt(fix2)/fix);

  float zip_result[][2]=zip(params, gradients);
  for(int i=0; i<sizeof(zip_result); i++){
    float p=zip_result[i][0];
    float g=zip_result[i][1];

    float mt=(this.beta*g)+(((float)1-this.beta)*p);
    float vt=(this.beta2*(g**2))+(((float)1-this.beta2)*g);
    float gt=mt/(sqrt(vt)+this.epsilon);
    float pt=p-(learning_rate_temp*gt);

    updates[i]={m, mt};
    updates[i+1]={v, vt};
    updates[i+2]={p, pt};
  }
  updates[i]={x, it};
  return updates;
}
struct* Adam::get_variables(){
  Adam_variables* vars;
  vars.iterations=this.iterations;
  vars.learning_rate=this.learning_rate;
  vars.beta=this.beta;
  vars.beta2=this.beta2;
  vars.init_decay=this.init_decay;
  vars.decay=this.decay;
  vars.epsilon=this.epsilon;
  return vars;
}

File: Simple-NN-library/python/functions/optimisers/adam.py
from further_maths import*

class Adam:
    def __init__(self, learning_rate=0.001, beta=0.9, beta2=0.999, decay=0):
        self.iterations=0
        self.learning_rate=learning_rate
        self.beta=beta
        self.beta2=beta2
        self.init_decay=decay
        self.decay=decay
        self.epsilon=epsilon()

    def update(self, cost, params=[]):
        updates=[]
        gradients=make_array(0, int(cost))
        x=0
        it=x+1
        fix=float(1)-(float(1)-self.beta)**it
        fix2=float(1)-(float(1)-self.beta2)**it
        learning_rate_temp=self.learning_rate*(sqrt(fix2)/fix)

        zip_result=zip(params, gradients)
        for i in range(len(zip_result)):
            p=zip_result[i][0]
            g=zip_result[i][1]

            mt=(self.beta*g)+((float(1)-self.beta)*p)
            vt=(self.beta2*(g**2))+((float(1)-self.beta2)*g)
            gt=mt/(sqrt(vt)+self.epsilon)
            pt=p-(learning_rate_temp*gt)

            updates.append((m, mt))
            updates.append((v, vt))
            updates.append((p, pt))
        updates.append((i, it))
        return updates

    def get_variables(self):
        return {
        'iterations':self.iterations,
        'learning_rate':self.learning_rate,
        'beta':self.beta, 'beta2':self.beta2,
        'init_decay':self.init_decay,
        'decay':self.decay,
        'epsilon':self.epsilon
        }
